File: SE463/Activities/RMA_Test/RMATest.java
import java.util.ArrayList;
import java.util.Scanner;
import java.util.TreeMap;
import java.util.TreeSet;

/**
 * This is the main class that implements the logic of all
 * three Rate Monotonic Analysis (RMA) schedulability tests.
 * 
 * @author Christopher A. Wood, caw4567@rit.edu
 */
public class RMATest
{
	/* Fixed upper bound for the first schedule test */
	final private double MAX_UPPER_BOUND = 1.0;
	
	/**
	 * TODO
	 * 
	 * @param tasks
	 * @return
	 */
	public boolean testOne(ArrayList<Task> tasks)
	{
		boolean result = false;
		
		// Compute the sum of C/P ratios among all tasks.
		double sum = sumUtilizationRatios(tasks);
		
		// Display the utilization for this task set.
		System.out.println("Test one calculated utilization: " + sum);
		
		// Now check the result 
		if (sum <= MAX_UPPER_BOUND)
		{
			result = true;
		}
		
		return result;
	}
	
	/**
	 * TODO
	 * 
	 * @param tasks
	 * @return
	 */
	public boolean testTwo(ArrayList<Task> tasks)
	{
		boolean result = false;
		
		// Compute the sum of C/P ratios and compare it to 1.0
		double sum = sumUtilizationRatios(tasks);
		
		// Display the utilization
		System.out.println("Test one calculated utilization: " + sum);
		
		// Compute the upper bound based on the number of tasks according to the following:
		// bound = n(2^(1/n) - 1)
		int n = tasks.size();
		double lowUpperBound = n * (Math.pow(2.0, (1/(double)n)) - 1);
		
		// Now check the result
		if (sum <= lowUpperBound)
		{
			result = true;
		}
		
		return result;
	}

	/**
	 * TODO
	 * 
	 * @param tasks
	 * @return
	 */
	public boolean testThree(ArrayList<Task> tasks)
	{
		boolean result = false;
		double sum = 0.0;
		int numTasks = 0;
		
		// Traverse the task set and apply Theorem 2 for each
		// task that does not pass Theorem 1
		for (Task t : tasks)
		{
			numTasks++;
			sum += t.utilizationRatio();
			double bound = numTasks * (Math.pow(2.0, (1/(double)numTasks)) - 1);
			
			// If we don't pass Theorem 1, apply Theorem 2
			if (sum < bound)
			{
				result = incrementalTest(tasks, numTasks - 1);
				if (!result)
				{
					System.out.println("Failed incremental application of Theorem 2");
					break;
				}
			}
		}
		
		return result;
	}
	
	/**
	 * Direct application of Theorem 2 from the paper.***
	 * 
	 * @param tasks
	 * @param index
	 * @return
	 */
	public boolean incrementalTest(ArrayList<Task> tasks, int failIndex)
	{
		boolean result = false;
		TreeMap<Integer, ArrayList<Integer>> schedulePoints = new TreeMap<Integer, ArrayList<Integer>>();
		
		// 1. calculate all schedule points
		// 2. construct inequalities for all schedule points
		// 3. check to see if one is satisfiable (if so, pass)
		
		// Determine task with the maximum period
		int maxPeriod = 0;
		for (int index = 0; index <= failIndex; index++)
		{
			if (tasks.get(index).getPeriod() > maxPeriod)
			{
				maxPeriod = tasks.get(index).getPeriod();
			}
		}
		
		// Determine schedule points for these tasks based on this lowest frequency task
		for (int index = 0; index <= failIndex; index++)
		{
			// Create the list of schedule points for this task
			ArrayList<Integer> points = new ArrayList<Integer>();
			
			// Compute all integer multiples of this task's period that are 
			// less than the maximum period for the subset of tasks
			int pointMult = 1;
			while (pointMult * tasks.get(index).getPeriod() <= maxPeriod)
			{
				points.add(tasks.get(index).getPeriod() * pointMult);
			}
			
			// Save this task's list of schedule points
			schedulePoints.put(index, points);
		}
		
		// Construct the scheduling inequalities (recursive approach)
		for (Integer base : schedulePoints.keySet())
		{
			for (int pointMult = 0; pointMult < schedulePoints.get(base).size(); pointMult++)
			{
				result = checkInequality(schedulePoints, base, pointMult);
				if (result == true)
				{
					System.out.println("one inequality satisfied!");
					break;
				}
			}
			
			// Break out of the parent loop
			if (result == true)
			{
				break;
			}
		}
		
		return result;
	}
	
	public boolean checkInequality(TreeMap<Integer, ArrayList<Integer>> points, int pointBase, int listOffset)
	{
		boolean result = false;
		int rhs = points.get(pointBase).get(listOffset);
		
		// The map that stores integer multiples of tasks when checking the inequalities
		TreeMap<Integer, Integer> multMap = new TreeMap<Integer, Integer>();
		for (Integer base : points.keySet())
		{
			multMap.put(base, 1); // give each task a multiplicity of 1 to start
		}
		
		// Compute all possible combinations of the tasks such that they are all less than the RHS
		
		
		return result;
	}
	
	/**
	 * Helper method that computes the sum of the utilization ratios for
	 * a given set of tasks.
	 * 
	 * @param tasks - a set of tasks
	 * @return the sum of the utilization ratios.
	 */
	public double sumUtilizationRatios(ArrayList<Task> tasks)
	{
		double sum = 0.0;
		
		// Traverse the task set and compute the ratios
		for (Task t : tasks)
		{
			sum += t.utilizationRatio();
		}
		
		return sum;
	}
	
	public static void main(String[] args)
	{
		int numTasks = 0;
		ArrayList<Task> tasks = new ArrayList<Task>();
		Scanner scin = new Scanner(System.in);
		
		// Read in the task parameters from stdin
		System.out.println("Number of tasks: ");
		try 
		{
			// Read in the number of tasks
			numTasks = scin.nextInt();
			assert(numTasks > 0);
			
			// Iteratively gather the remaining task data
			// TODO: change this to loop until EOF (ctl+d stdin)
			for (int i = 0; i < numTasks; i++)
			{
				System.out.println("Task " + i + " data (compute and period times separated by a space): ");
				int ct = scin.nextInt();
				int pt = scin.nextInt();
				
				// Verify these parameters 
				// Note: we assume compute time is strictly less than period time,
				// otherwise the schedulability would be trivially false with n>1 tasks
				assert(ct > 0 && pt > 0 && ct < pt); 
				// TODO: replace with simple check/stderr out?
				
				// Add the task to the running set
				tasks.add(new Task(ct, pt));
			}
			
			// If the task parameters were read in correctly, create the 
			// tester to run each scheduling test
			RMATest tester = new RMATest();
			
			// Perform the tests
			boolean testOnePass = tester.testOne(tasks);
			boolean testTwoPass = tester.testTwo(tasks);
			boolean testThreePass = tester.testThree(tasks);
			
			// Display the results
			System.out.println("Test one result: " + testOnePass);
			System.out.println("Test two result: " + testTwoPass);
			System.out.println("Test three result: " + testThreePass);
		}
		catch (Exception e)
		{
			System.err.println("Error parsing standard input.");
		}		
	}
}

File: SE463/Activities/RMA_Test/RMATest.cpp
//******************************************************************************
//
// File:    RMATest.cpp
// Author:  Christopher Wood, caw4567@rit.edu
// Version: 12/9/11
//
//******************************************************************************

// File includes
#include <iostream>
#include <stdlib.h>
#include <math.h>
#include <vector>
#include <cassert>

// For simplicity
using namespace std;

// Useful constants and definitions
#define RATIO_UPPER_BOUND (1)

// Store task data information
typedef struct
{
	int computeTime;
	int periodTime;
} TaskData;

/**
 * The first schedule test that simply checks to see if the 
 * utilization ratio sum is less than or equal to 1. 
 *
 * @param sum - the accumulated utilization ratio sum.
 *
 * @return true if the test passed, false otherwise.
 */
inline bool testOne(float sum)
{
	return sum <= RATIO_UPPER_BOUND;
}

/**
 * The second schedule test that checks to see if the 
 * utilization ratio sum is less than or equal to the following:
 *
 * n * (2^(1/n) - 1)
 *
 * @param sum - the accumulated utilization ratio sum.
 * @param numTasks - the number of tasks for this test.
 *
 * @return true if the test passed, false otherwise.
 */
inline bool testTwo(float sum, int numTasks)
{
	return sum <= ((float)numTasks * (pow(2, (1 / (float)numTasks)) - 1));
}

/**
 * The third schedule test that checks to see if there 
 * is some task ordering such that, if scheduled, will
 * allow each task at a specific instance in time to meet
 * its required deadline. 
 *
 * @param numTasks - the number of tasks for this test.
 * @param taskList - list of task data for each task.
 *
 * @return true if the test passed, false otherwise.
 */
bool testThree(int numTasks, vector<TaskData> taskList)
{
	float taskSum = 0.0;
	vector<bool> validList;
	bool result = true; // hope for the best
	
	// Traverse over the entire task set and perform the third test.
	for(int i = 1; i <= numTasks; i++)
	{
		// Assume invalid to start
		validList.push_back(false);
		
		// Calculate the k index for this run (which is used for the l index).
		for(int k = 1; k <= i; k++)
		{
			// Calculate the l index
			for (int l = 1; l <= floor((float)taskList[i].periodTime / (float)taskList[k].periodTime); l++)
			{
				// Continue on with the j index to compute the desired sum.
				taskSum = 0.0;
				for(int j = 1; j <= i; j++)
				{
					// This is the main calculation.
					taskSum += (((float)taskList[j].computeTime / (float)(l * taskList[k].periodTime)) * 
								ceil((float)(l * taskList[k].periodTime) / (float)taskList[j].periodTime));
				}
			
				// Check to see if the sum is less than the upper bound, as desired.
				if (taskSum <= RATIO_UPPER_BOUND)
				{
					validList[i - 1] = true; // flag this task schedule as valid
				}
			}
		}
	}
	
	// Check to make sure there is a valid task order for each task in the list.
	for (vector<bool>::iterator itr = validList.begin(); itr != validList.end(); itr++)
	{
		if (*itr == false)
		{
			result = false;
			break;
		}
	}
	
	return result;
}

/**
 * The main entry point of the program that reads in task parameters
 * and invokes the three schedule tests.
 */
int main(int argc, char *argv[])
{	
	int numTasks = 0;
	int ct = false;
	int pt = false;
	float utilSum = 0.0;
	vector<TaskData> taskList;
	
	// Retrieve the number of tasks
	cout << "Enter the number of tasks (n > 0): ";
	cin >> numTasks;
	assert(numTasks > 0);
	
	// Retrieve the parameters for each task, validating them
	// one at a time and storing them in the list of tasks
	cout << "Enter the details of each task in the following format" 
		<< endl << "\tCOMPUTE PERIOD" << endl;
	
	// Push back some dummy data to help with test three.
	for (int counter = 0; counter < numTasks; counter++)
	{
		// Read in a task tuple
		cin >> ct;
		cin >> pt;
		
		// Verify they are a correct/possible pair
		assert(ct > 0 && pt > 0 && ct <= pt);
		
		// Accumulate the utilization ratio sum
		utilSum += ((float)ct / (float)pt);
		
		// Store the resulting times in the two appropriate lists
		TaskData data;
		data.computeTime = ct;
		data.periodTime = pt;
		taskList.push_back(data);
	}
	
	// Sort the tasks based on their period time (bubble sort)
	for (int i = 0; i < taskList.size(); i++)
	{
		for (int j = 0; j < taskList.size() - 1; j++)
		{
			if (taskList[j].periodTime > taskList[j + 1].periodTime)
			{
				TaskData temp = taskList[j + 1];
				taskList[j + 1] = taskList[j];
				taskList[j] = temp;
			}
		}
	}
	
	// Insert a dummy value at the start of the list
	TaskData dummy;
	dummy.periodTime = 0;
	dummy.computeTime = 0;
	taskList.insert(taskList.begin(), dummy);
	
	// The first two tests are trivial, so run them first
	cout << "---------------------------------" << endl;
	cout << "Test results (0 - fail, 1 - pass)" << endl;
	cout << "TEST 1: " << testOne(utilSum) << endl;
	cout << "TEST 2: " << testTwo(utilSum, numTasks) << endl;
	cout << "Test 3: " << testThree(numTasks, taskList) << endl;
	cout << "---------------------------------" << endl;
	
	// Terminate gracefully.
	return 0;
}
   